Create scoped client

T0mats0 opened the pull request:
This is a pull request for issue #14127 . I read the discussion. I agree with Admad about not muddling up the constructor with mixed params. I also didn't like the idea of scoping the client after construction.  So I came up with the idea to make a static factory method that does the parsing for the string and returns a scoped client. I can see how this would be useful for folks storing string endpoints in their configuration or database. I thought about creating a factory class, but went with a method instead.

Let me know what you think.

Commit:
Created static factory method to create a scoped http client with a url string eliminating the need for calling code to parse such strings.

Commit:
Return type changed to self.

Commit:
Formatting fixes

Commit:
Formatting fixes

Commit:
Documentation, added exception. Removed unused variable.

dereuromark commented on src/Http/Client.php:
```diff
@@ -227,6 +227,30 @@ public function addCookie(CookieInterface $cookie)
+     * @param  string  $url A string url e.g. https://example.com
+     * @return self
+     */
+    public static function createScopedClientFromUrl(string $url): self
```
The documented type is `static`, and there should be no return typehint on these methods.

dereuromark commented on tests/TestCase/Http/ClientTest.php:
```diff
@@ -898,4 +898,61 @@ public function testRedirectDifferentSubDomains()
+    public function testCreateScopedClientSetsPort()
+    {
+        $client = Client::createScopedClientFromUrl('https://example.co:8765/');
+        $this->assertEquals('8765', $client->getConfig('port'));
```
We prefer assertSame() here to make sure the type is also as expected.

Commit:
Updated doc block and test

dereuromark commented on tests/TestCase/Http/ClientTest.php:
```diff
@@ -898,4 +898,61 @@ public function testRedirectDifferentSubDomains()
+    public function testCreateScopedClientSetsProtocol()
+    {
+        $client = Client::createScopedClientFromUrl('https://example.co/');
+        $this->assertEquals('https', $client->getConfig('scheme'));
```
There are a few more of those.

T0mats0 replied:
:+1:

Commit:
AssertSame not AssertEquals in more of the functions.

Commit:
Fully qualified name in documentation.

dereuromark commented on src/Http/Client.php:
```diff
@@ -227,6 +227,31 @@ public function addCookie(CookieInterface $cookie)
+     * Pass a string url to set the base domain and scheme for the request.
+     * Url must parse into a scheme and host.
+     *
+     * @param  string  $url A string url e.g. https://example.com
```
There are some extra white space in there that shouldn't be.
Also, in text URL is usually used as acronym.

Commit:
Patched up doc block

ADmad commented:
There should also be a test showing that only scheme, port and host are used. Any additional URL parts are ignored. The docblock could clarify the same too.

ADmad commented:
I wonder if we could add support for a `basePath` config too. So using URL `https://foo.com/api/` would result in `basePath` being `api/` and calling `get('bar')` would make a request to `https://foo.com/api/bar`.

Commit:
Throws InvalidArgumentException when the parsed string returns false instead of parts. Wrote test to make sure that the parsed url parts do not overwrite other defaults.

Commit:
URL in caps in all the docblocks.

T0mats0 commented:
> I wonder if we could add support for a `basePath` config too. So using URL `https://foo.com/api/` would result in `basePath` being `api/` and calling `->get('bar')` would make a request to `https://foo.com/api/bar`.

I like this idea it would make the the proposed function more useful if I'm using the class to interact with an api.

Commit:
Updated doc block.

T0mats0 commented on src/Http/Client.php:
```diff
@@ -227,6 +227,31 @@ public function addCookie(CookieInterface $cookie)
+        $parts = parse_url($url);
+
+        if ($parts === false) {
+            $parts = [];
```
Made this an exception, and I found a string that produces a boolean false response (see test). It's difficult to make the function return false ('', 'somestring, '123',  '_' all return an array).

T0mats0 commented on src/Http/Client.php:
```diff
@@ -227,6 +227,31 @@ public function addCookie(CookieInterface $cookie)
+        $config = array_intersect_key($parts, ['scheme' => '', 'port' => '', 'host' => '']);
+
+        if (!isset($config['scheme']) || !isset($config['host'])) {
+            throw new InvalidArgumentException('String ' . $url . ' did not parse');
```
Updated the error message.

markstory commented on tests/TestCase/Http/ClientTest.php:
```diff
@@ -898,4 +898,96 @@ public function testRedirectDifferentSubDomains()
+        try {
+            Client::createScopedClientFromUrl('https://');
+        } catch (InvalidArgumentException $e) {
+            $this->assertInstanceOf(InvalidArgumentException::class, $e);
```
You can use `expectException()` instead of try/catch.

T0mats0 replied:
Corrected

markstory commented on tests/TestCase/Http/ClientTest.php:
```diff
@@ -898,4 +898,96 @@ public function testRedirectDifferentSubDomains()
+    {
+        try {
+            Client::createScopedClientFromUrl('example.co');
+        } catch (InvalidArgumentException $e) {
```
You can use `expectException()` instead of try/catch.

T0mats0 replied:
Corrected

markstory commented on tests/TestCase/Http/ClientTest.php:
```diff
@@ -898,4 +898,96 @@ public function testRedirectDifferentSubDomains()
+    {
+        try {
+            Client::createScopedClientFromUrl('htps://');
+        } catch (InvalidArgumentException $e) {
```
You can use `expectException()` and `expectExceptionMessage()` instead of this try/catch.

T0mats0 replied:
Corrected.

ADmad commented:
> I like this idea it would make the the proposed function more useful if I'm using the class to interact with an api.

Yup, accessing APIs is probably the most common use case for the HTTP client and often APIs have a common path prefix. So adding this feature would avoid have to repeat the common path fragment for each request call.

Let's see what others have to say about it.

ADmad commented on src/Http/Client.php:
```diff
@@ -227,6 +227,32 @@ public function addCookie(CookieInterface $cookie)
+     * @return static
+     * @throws \InvalidArgumentException
+     */
+    public static function createScopedClientFromUrl(string $url)
```
I don't think using "scoped" in the name is necessary and the "client" is also redundant. The name could be shortened to `createFromUrl()`.

Commit:
Using expectException in tests.
Update method name to createFromUrl()

Commit:
Removed protocol in test name, used scheme instead.

T0mats0 commented:
I've updated the name of the method to createFromUrl. I've adjusted the tests to use expectsException().

Do we want to add 'basePath' (uri segment), to the method as proposed by Admad? If so I would propose adding it to the defaults and options that get passed around the rest of the class. 

I like the idea, it would make the class more useful for calling api's where there is a base path or uri segment after the domain.

Commit:
basePath configuration added to the client
tests updated to use basePath prepended and appended forward slash characters are stripped and re-added when building a url
Updated tests for exceptions and required variables

Commit:
Updated documentation

markstory commented:
I think you'll need to merge 4.next back into this branch or rebase to get CI passing.

ADmad commented:
@T0mats0 I think we can go ahead with the addition of "basePath" option. No one has objected to it's addition:).

T0mats0 commented:
Okay. I'll merge 4.x. My last push has basePath added.

Commit:
Merge remote-tracking branch 'upstream/4.next' into CreateScopedClient

T0mats0 commented on tests/TestCase/Http/ClientTest.php:
```diff
@@ -902,82 +902,69 @@ public function testRedirectDifferentSubDomains()
-        $this->fail('Client should have thrown exception');
+        $this->expectException(InvalidArgumentException::class);
+        Client::createFromUrl('https://');
+        $message = $this->getExpectedExceptionMessage();
```
It was not, but I've updated the code. It should be good now.

T0mats0 commented:
I fetched and merged 4.next. Some database errors are still occurring. I'm not sure why.

Commit:
Tests passing for adding basePath.

T0mats0 commented:
... and we're green again except for the static analyzer does not like "new static".

basePath option was added to the client class.

othercorey commented:
> ... and we're green again except for the static analyzer does not like "new static".

You need to add it to the baseline file so it is ignored. Here is an example https://github.com/cakephp/cakephp/blob/de98d012fbd258c77b774b891ddd3c6e37020e49/phpstan-baseline.neon#L119

I would also like to see this function move to just below the constructor.

Commit:
Moved function just below constructor, added ignore rule for phpstan for new static call.

Commit:
updated phpstan configuration to ignore new static in http client.

ADmad commented on tests/TestCase/Http/ClientTest.php:
```diff
@@ -898,4 +920,98 @@ public function testRedirectDifferentSubDomains()
+    /**
+     * basePath is set when passed to client in string
+     */
+    public function testCreateFromUrlSetsBasePath()
```
Instead of individual test methods like these you can add a provider method for URLs and then a single test methods which asserts the scheme, host, port and basePath generated.
